### ISSUE TITLE
Manually unrolling cuDNN RNN OOM

### ISSUE THREAD

**Issue by csarofeen** [pytorch/pytorch#914]
Manually unrolling cuDNN backend will cause memory usage to go sky high. 

Unrolled non-cuDNN pytorch takes ~1.8GB mem.
Non-unrolled cuDNN can take ~3GB mem.
Manually unrolling over time in user script will take >12GB mem.

Important for attention models @bmccann. Repro code below.

```.py
import torch
import torch.optim as optim
import torch.nn as nn
import time
from torch.autograd import Variable
from time import sleep

#torch.backends.cudnn.enabled=False
#torch.cuda.set_device(1)

input_size = 1024
hidden_size = 1024*2
batch_size = 64
time_steps = 32*2
lr   = 4e-2

def initHidden(bsz):
    return (Variable(torch.cuda.FloatTensor(2, bsz, hidden_size).zero_()),
            Variable(torch.cuda.FloatTensor(2, bsz, hidden_size).zero_()))
def resetHidden(hidden):
    return (Variable(hidden[0].data.zero_()), Variable(hidden[1].data.zero_()))

model = nn.LSTM(input_size, hidden_size, num_layers=2).cuda()
input = torch.randn(time_steps, batch_size, input_size).cuda()
target = torch.randn(time_steps, batch_size, hidden_size).cuda()


optimizer = optim.SGD(model.parameters(),
                      lr = lr,
                      momentum=0.9,
                      dampening = 0.0,
                      weight_decay = 0.0
                     )

criterion = nn.MSELoss().cuda()

loss = 0

hidden = initHidden(batch_size)
input = Variable(input)
target = Variable(target, requires_grad=False)

for i in range(1):
    start = time.time()
    for epoch in range(9999):
        print(epoch)
        loss = 0
        model.zero_grad()
        optimizer.zero_grad()

        hidden=initHidden(batch_size)
#        output, hidden = model(input, hidden)
        outputs = []
        for j in range(input.size(0)):
            output, hidden = model(input[j].view(1, *input[j].size()), hidden)
            outputs.append(output)
        outputs = torch.cat(outputs, 0)
        output = outputs

        
        loss = criterion(output, target)
        loss.backward(retain_variables = True)
        
        optimizer.step()

    print("Test ran in " + str( time.time() - start) + " seconds")
```

**Comment by apaszke** [comment]
What are unrolled, non-unrolled and manually unrolled RNNs? I don't really understand what you mean.

**Comment by csarofeen** [comment]
@apaszke Unrolling meaning sending a single step of input to the RNN at a time. i.e. manually unrolling the first dim of input (non-batch-first). This is important if you need to know something about the result of the previous step before you know what to send as input to the next step.

**Comment by apaszke** [comment]
Thanks. And how are these two different?

Unrolled non-cuDNN pytorch takes ~1.8GB mem.
Manually unrolling over time in user script will take >12GB mem.

**Comment by ngimel** [comment]
First one is not using cudnn, second one is. Both are unrolled in the same way.

**Comment by csarofeen** [comment]
```.py
output, hidden = model(input, hidden)
```
versus
```.py
outputs = []
for j in range(input.size(0)):
  output, hidden = model(input[j].view(1, *input[j].size()), hidden)
  outputs.append(output)
outputs = torch.cat(outputs, 0)
```

**Comment by apaszke** [comment]
Thanks :+1: I just wanted to be sure that I understand what exact case causes the problem

**Comment by apaszke** [comment]
One simple fix was to make the workspace local to the function call. Right now it has been cached in the function object and it reserved at least 250MB per each time step even when a very small batch size was used. Before the change I could only fit in memory a batch of 8 sequences of length 20, but after a batch size of 64 with 35 time steps can be used. But there must be something else too.

**Comment by apaszke** [comment]
It seems that the weight buffer takes up another 224 MB per time step.

**Comment by jekbradbury** [comment]
Is there any way to rearrange the `Storage`s on the first call to a particular `nn.LSTM` instance into the weight buffer that CUDNN expects,  rather than copying the weight buffer every time?

**Comment by apaszke** [comment]
The original script can run and use 2,5GB, if only we reuse bufs between time steps.

**Comment by apaszke** [comment]
@jekbradbury yeah we could flatten the params, but it's quite an ugly hack 😕

**Comment by ngimel** [comment]
Is there a way to make it less ugly?

**Comment by jekbradbury** [comment]
I'm in favor of flattening; I think copying buffers at runtime is Not Good and we don't really have a choice but to fit the wrapper as closely as possible to things CUDNN wants (as you did with the packed input arrays).
BTW Chainer's `functions.n_step_lstm` copies buffers at runtime, while cudnn.torch and tf.contrib.cudnn_rnn flatten params and allow non-flattened access if the user wants to pull the params out for something else.

**Comment by apaszke** [comment]
I can't think of anything else right now. The problem is that we'd like to keep the modules as stateless as possible, and to isolate them from functions they use. I'd like to limit the leaks that way. I'll try to come up with something else, and if there's no other solution, we'll have to add the flattening.

**Comment by jekbradbury** [comment]
I guess the idea is that parameters are Variables, so Functions should be able to silently move their underlying Tensors (after first checking if they're already in acceptable memory positions). That would break any user code that assumes that the Tensors underlying a network's parameters will always stay in the same place in memory -- but that code would already be broken by someone calling .cuda(another_gpu) on the model.

**Comment by apaszke** [comment]
I don't really like it, because it's quite magical. However, I think it would be possible to keep a class dict in CudnnRNN Function, that would keep the weight buffers for reuse. Once the last function referencing the buffer quits, it would get deleted too. Additionally, it would register a hook on the weight Variables, to ensure that changes to parameters would invalidate the cache.

**Comment by ngimel** [comment]
DataParallel already remaps parameters of all the modules, doing the same thing for RNNs is not that different (of course, flattened RNN parameters will break after dataParallel broadcasts them, but it will happen once per batch, not once per timestep, as it does now).

**Comment by ayushgithub** [comment]
Any update on this issue. I have a quite complex attention model and need to manually unroll the lstm. After switching off cudnn backend, my iteration takes 3-4 secs which will take too long to train. I do not want to move back to torch :P

**Comment by ritheshkumar95** [comment]
+1, i raised the same issue over here:
https://discuss.pytorch.org/t/rnn-open-loop-mode-oom/1820
